lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,484 @@
+
+    if idx and l then
+        l:set(idx, widget2)
+        return true
+    end
+
+    return false
+end
+
+function grid:swap_widgets(widget1, widget2, recursive)
+    base.check_widget(widget1)
+    base.check_widget(widget2)
+
+    local idx1, l1 = self:index(widget1, recursive)
+    local idx2, l2 = self:index(widget2, recursive)
+
+    if idx1 and l1 and idx2 and l2 and (l1.set or l1.set_widget) and (l2.set or l2.set_widget) then
+        if l1.set then
+            l1:set(idx1, widget2)
+        elseif l1.set_widget then
+            l1:set_widget(widget2)
+        end
+        if l2.set then
+            l2:set(idx2, widget1)
+        elseif l2.set_widget then
+            l2:set_widget(widget1)
+        end
+
+        return true
+    end
+
+    return false
+end
+
+function grid:set(index, widget2)
+    if (not widget2) or (not self._private.rows[index]) then return false end
+
+    base.check_widget(widget2)
+
+    self._private.rows[index] = widget2
+
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+--- Insert a new widget in the layout at position `index`
+-- @tparam number index The position
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+function grid:insert(index, widget)
+    local TODO_TOTAL = math.huge --FIXME
+    if not index or index < 1 or index > TODO_TOTAL then return false end
+
+    base.check_widget(widget)
+    table.insert(self._private.rows, index, widget) --FIXME
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+-- Fit the grid layout into the given space
+-- @param context The context in which we are fit.
+-- @param orig_width The available width.
+-- @param orig_height The available height.
+function grid:fit(context, orig_width, orig_height)
+    local max_col, max_row = get_cell_sizes(self, context, orig_width, orig_height)
+
+    -- Now that all fit is done, get the maximum
+    local used_max_h, used_max_w = 0, 0
+
+    for row_idx = 1, #max_row do
+        -- The other widgets will be discarded
+        if used_max_h + max_row[row_idx] > orig_height then
+            break
+        end
+
+        used_max_h = used_max_h + max_row[row_idx]
+    end
+
+    for col_idx = 1, #max_col do
+        -- The other widgets will be discarded
+        if used_max_w + max_col[col_idx] > orig_width then
+            break
+        end
+
+        used_max_w = used_max_w + max_col[col_idx]
+    end
+
+    return used_max_w, used_max_h
+end
+
+function grid:reset()
+    self._private.rows = {}
+    self._private.children = {}
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Set the layout's fill_space property. If true, the remaining space will
+-- be distributed equally across all widgets in the grid.
+-- @DOC_wibox_layout_grid_fill_space_EXAMPLE@
+-- @property fill_space
+-- @param boolean
+
+function grid:set_fill_space(val)
+    if self._private.fill_space ~= val then
+        self._private.fill_space = not not val
```

Comment:
Shouldn't you do the "not not dance" before the `~=`? Not that it matters much, just saying...
So something like:
```
val = not not val
if self._private.fill_space == val then return end
self._private.fill_space = val
etc
```